File: packages/ui-mobile-base/windows/widgets/ImageHelper.cs
using System;
using System.Runtime.InteropServices.WindowsRuntime;
using System.Threading.Tasks;
using Windows.Foundation;
using Windows.Graphics.Imaging;
using Windows.Storage;
using Windows.Storage.Streams;
using Windows.UI.Xaml.Media.Imaging;
using Windows.Web.Http;
using Windows.ApplicationModel.Core;
using Windows.UI.Core;

namespace NativeScript.Widgets
{
    public sealed class ImageResult
    {
        internal ImageResult(BitmapImage bitmap, IBuffer rawBuffer, int width, int height)
        {
            Bitmap    = bitmap;
            RawBuffer = rawBuffer;
            Width     = width;
            Height    = height;
        }

        public BitmapImage Bitmap    { get; }
        public IBuffer     RawBuffer { get; }
        public int         Width     { get; }
        public int         Height    { get; }
    }

    public sealed class ImageHelper
    {
        // ── Public API ────────────────────────────────────────────────────────

        public static IAsyncOperation<ImageResult> LoadFromBufferAsync(IBuffer buffer)
            => LoadFromBufferInternalAsync(buffer).AsAsyncOperation();

        public static IAsyncOperation<ImageResult> LoadFromUrlAsync(string url)
            => LoadFromUrlInternalAsync(url).AsAsyncOperation();

        public static IAsyncOperation<ImageResult> LoadFromFileAsync(string path)
            => LoadFromFileInternalAsync(path).AsAsyncOperation();

        public static IAsyncOperation<ImageResult> ResizeAsync(IBuffer buffer, int maxSize)
            => ResizeInternalAsync(buffer, maxSize).AsAsyncOperation();

        public static IAsyncOperation<bool> SaveToFileAsync(IBuffer buffer, string path)
            => SaveToFileInternalAsync(buffer, path).AsAsyncOperation();

        // Single native WinRT call — returns IAsyncOperation<IBuffer> directly.
        public static IAsyncOperation<IBuffer> ReadFileAsync(string path)
            => PathIO.ReadBufferAsync(path);

        public static string BufferToBase64(IBuffer buffer)
            => Convert.ToBase64String(buffer.ToArray());

        // ── Private async implementations ────────────────────────────────────

        private static async Task<ImageResult> LoadFromBufferInternalAsync(IBuffer buffer)
        {
            var stream = await WriteToStreamAsync(buffer);
            var bmp    = await BitmapFromStreamAsync(stream);
            return new ImageResult(bmp, buffer, bmp.PixelWidth, bmp.PixelHeight);
        }

        private static async Task<ImageResult> LoadFromUrlInternalAsync(string url)
        {
            var client = new HttpClient();
            var buffer = await client.GetBufferAsync(new Uri(url));
            var stream = await WriteToStreamAsync(buffer);
            var bmp    = await BitmapFromStreamAsync(stream);
            return new ImageResult(bmp, buffer, bmp.PixelWidth, bmp.PixelHeight);
        }

        private static async Task<ImageResult> LoadFromFileInternalAsync(string path)
        {
            var buffer = await PathIO.ReadBufferAsync(path);
            var stream = await WriteToStreamAsync(buffer);
            var bmp    = await BitmapFromStreamAsync(stream);
            return new ImageResult(bmp, buffer, bmp.PixelWidth, bmp.PixelHeight);
        }

        private static async Task<ImageResult> ResizeInternalAsync(IBuffer buffer, int maxSize)
        {
            var inStream = await WriteToStreamAsync(buffer);
            var decoder  = await BitmapDecoder.CreateAsync(inStream);

            uint srcW  = decoder.PixelWidth;
            uint srcH  = decoder.PixelHeight;
            double scale = (double)maxSize / Math.Max(srcW, srcH);
            uint dstW  = (uint)Math.Round(srcW * scale);
            uint dstH  = (uint)Math.Round(srcH * scale);

            var outStream = new InMemoryRandomAccessStream();
            var encoder   = await BitmapEncoder.CreateForTranscodingAsync(outStream, decoder);
            encoder.BitmapTransform.ScaledWidth  = dstW;
            encoder.BitmapTransform.ScaledHeight = dstH;
            await encoder.FlushAsync();

            outStream.Seek(0);
            var bmp = await BitmapFromStreamAsync(outStream);

            outStream.Seek(0);
            var reader = new DataReader(outStream);
            await reader.LoadAsync((uint)outStream.Size);
            var outBuffer = reader.DetachBuffer();

            return new ImageResult(bmp, outBuffer, (int)dstW, (int)dstH);
        }

        private static async Task<bool> SaveToFileInternalAsync(IBuffer buffer, string path)
        {
            await PathIO.WriteBufferAsync(path, buffer);
            return true;
        }

        // ── Stream utilities ─────────────────────────────────────────────────

        private static async Task<InMemoryRandomAccessStream> WriteToStreamAsync(IBuffer buffer)
        {
            var stream = new InMemoryRandomAccessStream();
            await stream.WriteAsync(buffer);
            stream.Seek(0);
            return stream;
        }

        private static async Task<BitmapImage> BitmapFromStreamAsync(IRandomAccessStream stream)
        {
            // Ensure the stream is positioned at the start.
            try { stream.Seek(0); } catch { }

            // Prefer creating and setting the BitmapImage on the UI dispatcher
            // because XAML BitmapImage is an object that must be touched from
            // the UI thread on many Windows platforms.
            var dispatcher = CoreApplication.MainView?.CoreWindow?.Dispatcher;
            if (dispatcher != null)
            {
                var tcs = new TaskCompletionSource<BitmapImage>();

                try
                {
                    await dispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
                    {
                        try
                        {
                            var bmp = new BitmapImage();
                            // SetSourceAsync returns IAsyncAction; convert it to a Task
                            // and continue to set the TaskCompletionSource when done.
                            var action = bmp.SetSourceAsync(stream);
                            var task = action.AsTask();
                            task.ContinueWith(t =>
                            {
                                if (t.IsFaulted) tcs.TrySetException(t.Exception?.InnerException ?? t.Exception);
                                else if (t.IsCanceled) tcs.TrySetCanceled();
                                else tcs.TrySetResult(bmp);
                            }, TaskScheduler.Default);
                        }
                        catch (Exception ex)
                        {
                            tcs.TrySetException(ex);
                        }
                    });
                }
                catch
                {
                    // Dispatcher invocation failed; fall back to attempting creation
                    // on the current thread below.
                }

                return await tcs.Task;
            }

            // Fallback: try creating on current thread. This may fail if called
            // from a non-UI thread (caller should prefer the async API to be
            // invoked from appropriate context), but we catch and rethrow so
            // callers can handle errors.
            var fallbackBmp = new BitmapImage();
            await fallbackBmp.SetSourceAsync(stream);
            return fallbackBmp;
        }
    }
}

File: packages/ui-mobile-base/windows/widgets-cpp/ImageHelper.cpp
#include "pch.h"
#include "ImageResult.h"
#include "ImageHelper.h"
#include "ImageResult.g.cpp"
#include "ImageHelper.g.cpp"

#include <algorithm>
#include <cmath>

using namespace winrt;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::Storage;
using namespace winrt::Windows::Storage::Streams;
using namespace winrt::Windows::Graphics::Imaging;
using BitmapImage = winrt::Microsoft::UI::Xaml::Media::Imaging::BitmapImage;

// THREADING: these deliberately do NOT resume_background(). The NativeScript Windows
// runtime hosts V8 on the UI thread and only marshals WinRT async `Completed` callbacks back into
// JS when they fire on that (UI) apartment. WinRT async operations started on the UI apartment
// marshal their completion back to it automatically, so awaiting them inline keeps every resume on
// the UI thread — which is also where the UI-affine BitmapImage must be created. Hopping to a
// background thread (resume_background) would make the final completion fire on a pool thread where
// the runtime's thread-local isolate pointer is null, so the JS promise would silently never settle.
// The awaited ops are themselves async (non-blocking), so the UI thread is not stalled.

namespace
{
    // Copy an IBuffer into a fresh in-memory stream, rewound to the start.
    IAsyncOperation<InMemoryRandomAccessStream> WriteToStreamAsync(IBuffer buffer)
    {
        InMemoryRandomAccessStream stream;
        co_await stream.WriteAsync(buffer);
        stream.Seek(0);
        co_return stream;
    }

    IAsyncOperation<BitmapImage> BitmapFromStreamAsync(IRandomAccessStream stream)
    {
        try { stream.Seek(0); }
        catch (...) {}
        BitmapImage bmp;
        co_await bmp.SetSourceAsync(stream);
        co_return bmp;
    }
}

namespace winrt::NativeScript::Widgets::implementation
{
    IAsyncOperation<winrt::NativeScript::Widgets::ImageResult> ImageHelper::LoadFromBufferAsync(IBuffer buffer)
    {
        auto stream = co_await WriteToStreamAsync(buffer);
        auto bmp = co_await BitmapFromStreamAsync(stream);
        co_return winrt::make<ImageResult>(bmp, buffer, bmp.PixelWidth(), bmp.PixelHeight());
    }

    IAsyncOperation<winrt::NativeScript::Widgets::ImageResult> ImageHelper::LoadFromUrlAsync(hstring url)
    {
        winrt::Windows::Web::Http::HttpClient client;
        auto buffer = co_await client.GetBufferAsync(Uri{ url });
        auto stream = co_await WriteToStreamAsync(buffer);
        auto bmp = co_await BitmapFromStreamAsync(stream);
        co_return winrt::make<ImageResult>(bmp, buffer, bmp.PixelWidth(), bmp.PixelHeight());
    }

    IAsyncOperation<winrt::NativeScript::Widgets::ImageResult> ImageHelper::LoadFromFileAsync(hstring path)
    {
        auto buffer = co_await PathIO::ReadBufferAsync(path);
        auto stream = co_await WriteToStreamAsync(buffer);
        auto bmp = co_await BitmapFromStreamAsync(stream);
        co_return winrt::make<ImageResult>(bmp, buffer, bmp.PixelWidth(), bmp.PixelHeight());
    }

    IAsyncOperation<winrt::NativeScript::Widgets::ImageResult> ImageHelper::ResizeAsync(IBuffer buffer, int32_t maxSize)
    {
        auto inStream = co_await WriteToStreamAsync(buffer);
        auto decoder = co_await BitmapDecoder::CreateAsync(inStream);

        const uint32_t srcW = decoder.PixelWidth();
        const uint32_t srcH = decoder.PixelHeight();
        const double scale = static_cast<double>(maxSize) / (std::max)(srcW, srcH);
        const uint32_t dstW = static_cast<uint32_t>(std::lround(srcW * scale));
        const uint32_t dstH = static_cast<uint32_t>(std::lround(srcH * scale));

        InMemoryRandomAccessStream outStream;
        auto encoder = co_await BitmapEncoder::CreateForTranscodingAsync(outStream, decoder);
        encoder.BitmapTransform().ScaledWidth(dstW);
        encoder.BitmapTransform().ScaledHeight(dstH);
        co_await encoder.FlushAsync();

        outStream.Seek(0);
        auto bmp = co_await BitmapFromStreamAsync(outStream);

        outStream.Seek(0);
        DataReader reader{ outStream };
        co_await reader.LoadAsync(static_cast<uint32_t>(outStream.Size()));
        auto outBuffer = reader.DetachBuffer();

        co_return winrt::make<ImageResult>(bmp, outBuffer, static_cast<int32_t>(dstW), static_cast<int32_t>(dstH));
    }

    IAsyncOperation<bool> ImageHelper::SaveToFileAsync(IBuffer buffer, hstring path)
    {
        co_await PathIO::WriteBufferAsync(path, buffer);
        co_return true;
    }

    IAsyncOperation<IBuffer> ImageHelper::ReadFileAsync(hstring path)
    {
        // Single native call — return the operation directly.
        return PathIO::ReadBufferAsync(path);
    }

    hstring ImageHelper::BufferToBase64(IBuffer buffer)
    {
        return winrt::Windows::Security::Cryptography::CryptographicBuffer::EncodeToBase64String(buffer);
    }
}
